core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterManager.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.operator.exchange;
+
+import io.airlift.log.Logger;
+import io.airlift.units.DataSize;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class ScaleWriterManager
+{
+    private static final Logger log = Logger.get(ScaleWriterExchanger.class);
+
+    private final LocalExchangeMemoryManager memoryManager;
+    private final int totalWriters;
+    private final long bufferSizeMinScalingThreshold;
+    private final Supplier<Long> physicalWrittenBytesSupplier;
+    private final long writerMinSize;
+    private final Runnable setWritersFullyScaled;
+    private final AtomicBoolean isSetWritersFullyScaledExecuted = new AtomicBoolean(false);
+
+    // start with single writer and increase the writer count based on
+    // physical written bytes and buffer utilization.
+    @GuardedBy("this")
+    private int writerCount = 1;
+
+    @GuardedBy("this")
+    private long lastScaleUpPhysicalWrittenBytes;
```

Comment:
Both these are being accessed outside of lock even though they're declared `@GuardedBy("this")`.
I can see IDE warnings from this, please check if you have concurrency related inspection turned on.